packages/raystack/calendar/date-picker.tsx

```diff
@@ -2,71 +2,162 @@ import { Popover } from "~/popover";
+    setCalendarVal(day);
+    setInputState(undefined);
+    removeEventListeners();
   };
 
   function onDropdownOpen() {
-    isDropdownOpenedRef.current = true;
+    isDropdownOpenRef.current = true;
   }
 
   function onOpenChange(open?: boolean) {
-    if (!isDropdownOpenedRef.current) {
+    if (!isDropdownOpenRef.current && !(isInputFieldFocused.current && showCalendar)) {
       setShowCalendar(Boolean(open));
     }
 
-    isDropdownOpenedRef.current = false;
+    isDropdownOpenRef.current = false;
+  }
+
+  function handleInputFocus() {
+    if (isInputFieldFocused.current) return;
+    if (!showCalendar) setShowCalendar(true);
+  }
+
+  function handleInputBlur(event: React.FocusEvent) {
+    if (isInputFieldFocused.current) {
+        const el = event.relatedTarget as HTMLElement | null;
+        if (el && isElementOutside(el)) removeEventListeners();
+    } else {
+        registerEventListeners();
+        setTimeout(() => textFieldRef.current?.focus());
+    }
+  }
+
+  function handleKeyUp(event: React.KeyboardEvent) {
+    if (event.code === 'Enter' && textFieldRef.current) {
+        textFieldRef.current.blur();
+        removeEventListeners();
+    }
+  }
+
+  function handleInputChange(event: React.ChangeEvent<HTMLInputElement>) {
+    const { value } = (event.target);
+
+    const format = value.includes("/") ? "DD/MM/YYYY" : value.includes("-") ? "DD-MM-YYYY" : undefined;
+    const date = dayjs(value, format);
+
+    const isValidDate = date.isValid();
+
+    const isAfter = calendarProps?.startMonth !== undefined ? dayjs(date).isSameOrAfter(calendarProps.startMonth) : true;
+    const isBefore = calendarProps?.endMonth !== undefined ? dayjs(date).isSameOrBefore(calendarProps.endMonth) : true;
+
+    const isValid = isValidDate && isAfter && isBefore && dayjs(date).isSameOrBefore(dayjs());
+
+    if (isValid) {
+        setCalendarVal(date.toDate());
+        if (inputState === 'invalid') setInputState(undefined);
+    } else {
+        setInputState('invalid');
+    }
   }
 
   return (
     <Popover open={showCalendar} onOpenChange={onOpenChange}>
-      <Popover.Trigger asChild>
-        <TextField
-          value={dateValue}
-          trailing={<CalendarIcon />}
-          className={styles.datePickerInput}
-          readOnly
+      <TextField
+          ref={textFieldRef}
+          defaultValue={defaultDate}
+          trailing={<Popover.Trigger asChild><CalendarIcon /></Popover.Trigger>}
+          onChange={handleInputChange}
+          onFocus={handleInputFocus}
+          onBlur={handleInputBlur}
+          state={inputState}
+          placeholder={placeholder}
+          onKeyUp={handleKeyUp}
           {...textFieldProps}
-        />
-      </Popover.Trigger>
-      <Popover.Content side={side} className={styles.calendarPopover}>
+      />
+
+      <Popover.Content side={side} className={styles.calendarPopover} ref={contentRef}>
         <Calendar
           required={true}
           {...calendarProps}
```

Comment:
{...calendarProps} - We can make this the last prop.

Reply:
We don't want user to overwrite the fields that are below {...calendarProps}